### PR TITLE
Rpc dev

### PR DESCRIPTION
**Pull request by gtygo**
create and fix some grpc function like  createrawtx  &  sendrawtx

### CODE REVIEW

**Comment by gtygo** [comment]
CreateRawTx,sendRawTx and other Rpc interface

File: commands/box/ctl/root.go [hunk 1]
```diff
@@ -20,6 +17,9 @@ import (
 	"github.com/BOXFoundation/boxd/wallet"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"path"
+	"strconv"
+	"strings"
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: should be at the first section

File: commands/box/ctl/root.go [hunk 2]
```diff
@@ -337,6 +331,90 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
 	}
 }
 
+func decoderawtx(cmd *cobra.Command, args []string) {
+	fmt.Println("decoderawtx called")
+	if len(args) < 1 {
+		fmt.Println("")
+		return
+	}
+	tx_byte, err := hex.DecodeString(args[0])
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: camel name rule

File: commands/box/ctl/root.go [hunk 3]
```diff
@@ -337,6 +331,90 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+	for _, x := range txid_str {
+		tmp := crypto.HashType{}
+		tmp.SetString(x)
+		txid = append(txid, tmp)
+	}
+
+	vout_str := strings.Split(args[2], ",")
+	vout := make([]uint32, 0)
+	for _, x := range vout_str {
+		tmp, _ := strconv.Atoi(x)
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: don't set error returned as blank

File: commands/box/ctl/root.go [hunk 4]
```diff
@@ -337,6 +331,90 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+	}
+
+	to_str := strings.Split(args[3], ",")
+	to := make([]string, 0)
+	for _, x := range to_str {
+		to = append(to, x)
+	}
+
+	amount_str := strings.Split(args[4], ",")
+	amount := make([]uint64, 0)
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: better to name amounts

File: commands/box/ctl/root.go [hunk 5]
```diff
@@ -337,6 +331,90 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+	for _, x := range amount_str {
+		tmp, _ := strconv.Atoi(x)
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: as above

File: commands/box/ctl/root.go [hunk 6]
```diff
@@ -337,6 +331,90 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
 	}
 }
 
+func decoderawtx(cmd *cobra.Command, args []string) {
+	fmt.Println("decoderawtx called")
+	if len(args) < 1 {
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: len(args) == 1

File: commands/box/ctl/root.go [hunk 7]
```diff
@@ -337,6 +331,90 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+		fmt.Println("")
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: ???
error detail and usage here

File: rpc/rpcutil/transaction.go [hunk 8]
```diff
@@ -401,6 +404,31 @@ func MakeUnsignedTx(
...
+		uw := txlogic.NewUtxoWrap(from, height, total)
+		utxo := txlogic.MakePbUtxo(op, uw)
+		utxos = append(utxos, utxo)
+	}
+	changeAmt, overflowed := calcChangeAmount(amounts, 0, utxos...)
+	if overflowed {
+		return nil, nil, txlogic.ErrInsufficientBalance
+	}
+	tx, err := txlogic.MakeUnsignedTx(from, to, amounts, changeAmt, utxos...)
+	return tx, utxos, err
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: why return utxos??

File: commands/box/ctl/root.go [hunk 9]
```diff
@@ -337,6 +331,90 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+	mashal_tx, err := tx.Marshal()
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	str_tx := hex.EncodeToString(mashal_tx)
+	fmt.Println(str_tx)
+}
+
 func signMessageCmdFunc(cmd *cobra.Command, args []string) {
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: could implement it.

File: commands/box/ctl/root.go [hunk 10]
```diff
@@ -382,6 +460,33 @@ func validateMessageCmdFunc(cmd *cobra.Command, args []string) {
 		fmt.Println(args[0], " is a valid address")
 	}
 }
+func sendrawtx(cmd *cobra.Command, args []string) {
+	fmt.Println("sendrawtx called")
+	if len(args) < 1 {
+		fmt.Println("Invalide argument number")
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: error detail and usage

File: commands/box/ctl/root.go [hunk 11]
```diff
@@ -382,6 +460,33 @@ func validateMessageCmdFunc(cmd *cobra.Command, args []string) {
...
+		return
+	}
+	conn, err := rpcutil.GetGRPCConn(getRPCAddr())
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	defer conn.Close()
+	tx_byte, err := hex.DecodeString(args[0])
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: error handle

File: commands/box/ctl/root.go [hunk 12]
```diff
@@ -382,6 +460,33 @@ func validateMessageCmdFunc(cmd *cobra.Command, args []string) {
...
+	tx := types.Transaction{}
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: new

File: commands/box/ctl/root.go [hunk 13]
```diff
@@ -382,6 +460,33 @@ func validateMessageCmdFunc(cmd *cobra.Command, args []string) {
...
+	error := tx.Unmarshal(tx_byte)
+	if error != nil {
+		fmt.Println(error)
+		return
+	}
+	resp, err := rpcutil.SendTransaction(conn, &tx)
+	if err != nil {
+		fmt.Println(err)
+		return
+	} else {
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: make may report error.
no need else clause.

File: commands/box/token/root.go [hunk 14]
```diff
@@ -79,6 +80,14 @@ func createTokenCmdFunc(cmd *cobra.Command, args []string) {
 	tokenSymbol := args[3]
 	tokenTotalSupply, err2 := strconv.Atoi(args[4])
 	tokenDecimals, err3 := strconv.Atoi(args[5])
+	if tokenDecimals < 0 || tokenDecimals > 8 {
+		fmt.Println("Invalid argument format")
```
Comment (Jerick26) [inline commands/box/token/root.go]: more error detail

File: commands/box/token/root.go [hunk 15]
```diff
@@ -79,6 +80,14 @@ func createTokenCmdFunc(cmd *cobra.Command, args []string) {
...
+		return
+	}
+	if uint64(tokenTotalSupply) > math.MaxUint64/uint64(math.Pow10(tokenDecimals)) {
+		fmt.Println("Invalid argument format")
```
Comment (Jerick26) [inline commands/box/token/root.go]: more error detail

File: commands/box/wallet/root.go [hunk 16]
```diff
@@ -155,6 +153,38 @@ func listAccountCmdFunc(cmd *cobra.Command, args []string) {
...
+		fmt.Println(err)
+		return
+	}
+	walletinfo, exist := wltMgr.GetAccount(addr)
+	fmt.Println(exist)
+	if exist {
+		fmt.Println("path: ", walletinfo.Path)
+		fmt.Println("Address: ", walletinfo.Addr())
+		fmt.Printf("Pubkey Hash: %x \n", walletinfo.PubKeyHash())
+		fmt.Println("PrivKey: ", privkey)
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: no need to show private key

File: commands/box/wallet/root.go [hunk 17]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
 	}
 	fmt.Printf("Address: %s\nPrivate Key: %s\n", addr, privateKey)
 }
+func dumpwallet(cmd *cobra.Command, args []string) {
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: add blank line between two functions

File: commands/box/wallet/root.go [hunk 18]
```diff
@@ -155,6 +153,38 @@ func listAccountCmdFunc(cmd *cobra.Command, args []string) {
...
+	walletinfo, exist := wltMgr.GetAccount(addr)
+	fmt.Println(exist)
+	if exist {
+		fmt.Println("path: ", walletinfo.Path)
+		fmt.Println("Address: ", walletinfo.Addr())
+		fmt.Printf("Pubkey Hash: %x \n", walletinfo.PubKeyHash())
+		fmt.Println("PrivKey: ", privkey)
+		fmt.Println("UnLocked: ", walletinfo.Unlocked)
+	}
+}
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: add blank line between two functions

File: commands/box/wallet/root.go [hunk 19]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
 	}
 	fmt.Printf("Address: %s\nPrivate Key: %s\n", addr, privateKey)
 }
+func dumpwallet(cmd *cobra.Command, args []string) {
+	fmt.Println("dumpwallet called")
+	if len(args) < 1 {
+		fmt.Println("file name needed")
+		return
+	}
+	if len(args[0]) <= 4 {
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: ??

File: commands/box/wallet/root.go [hunk 20]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+		fmt.Println("file name error")
+		return
+	}
+	if args[0][len(args[0])-4:] != ".txt" {
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: why need .txt

File: commands/box/wallet/root.go [hunk 21]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+		fmt.Println("Input file type must be in txt file format")
+		return
+	}
+	_, err := os.Stat(args[0])
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: wrong check

File: commands/box/wallet/root.go [hunk 22]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+			fmt.Printf(" You have %d remaining input opportunities\n", i)
+			passphrase, err := wallet.ReadPassphraseStdin()
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+			privateKey, err := wltMgr.DumpPrivKey(addr, passphrase)
+			if err != nil {
+				fmt.Print("Wrong password ")
+			} else {
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: no need else clause

File: commands/box/wallet/root.go [hunk 23]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+		}
+		fmt.Print("addr: " + addr)
+		for i := 3; i >= 1; i-- {
+			fmt.Printf(" You have %d remaining input opportunities\n", i)
+			passphrase, err := wallet.ReadPassphraseStdin()
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+			privateKey, err := wltMgr.DumpPrivKey(addr, passphrase)
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: error may be others

File: commands/box/wallet/root.go [hunk 24]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+	success_num := 0
+	failed_num := 0
+	for _, x := range addrs {
+		addr := x
+		wltMgr, err := wallet.NewWalletManager(walletDir)
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+		fmt.Print("addr: " + addr)
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: output format!!!

File: commands/box/wallet/root.go [hunk 25]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+		for i := 3; i >= 1; i-- {
+			fmt.Printf(" You have %d remaining input opportunities\n", i)
+			passphrase, err := wallet.ReadPassphraseStdin()
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+			privateKey, err := wltMgr.DumpPrivKey(addr, passphrase)
+			if err != nil {
+				fmt.Print("Wrong password ")
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: output format

File: commands/box/wallet/root.go [hunk 26]
```diff
@@ -179,3 +209,72 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+	wltMgr, err := wallet.NewWalletManager(walletDir)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	for _, acc := range wltMgr.ListAccounts() {
+		addrs = append(addrs, acc.Addr())
+	}
+	success_num := 0
+	failed_num := 0
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: why need two variable, one is enough

File: rpc/pb/transaction.proto [hunk 27]
```diff
@@ -11,77 +11,106 @@ import "common.proto";
...
-    rpc MakeUnsignedTokenIssueTx(MakeTokenIssueTxReq) returns (MakeTokenIssueTxResp) {
+    rpc MakeUnsignedTokenIssueTx (MakeTokenIssueTxReq) returns (MakeTokenIssueTxResp) {
         option (google.api.http) = {
             post: "/v1/tx/makeunsignedtx/token/issue"
             body: "*"
         };
     }
-
-    rpc MakeUnsignedTokenTransferTx(MakeTokenTransferTxReq) returns (MakeTxResp) {
+    rpc SendRawTransaction (SendRawTransactionReq) returns (SendTransactionResp) {
```
Comment (Jerick26) [inline rpc/pb/transaction.proto]: blank line

File: rpc/pb/transaction.proto [hunk 28]
```diff
@@ -11,77 +11,106 @@ import "common.proto";
...
+        option (google.api.http) = {
+            post: "/v1/tx/sendrawtransaction"
+            body: "*"
+        };
+    }
+    rpc CreateRawTransaction (CreateRawTransactionReq) returns (CreateRawTransactionResp) {
```
Comment (Jerick26) [inline rpc/pb/transaction.proto]: no need to add this roc api

File: rpc/rpcutil/transaction.go [hunk 29]
```diff
@@ -401,6 +404,31 @@ func MakeUnsignedTx(
...
+) (*types.Transaction, []*rpcpb.Utxo, error) {
+	total := uint64(0)
+	for _, a := range amounts {
+		total += a
+	}
+
+	utxos := make([]*rpcpb.Utxo, 0)
+	for i := 0; i < len(txid); i++ {
+		hash := txid[i]
+		op := types.NewOutPoint(&hash, vout[i])
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: use txHash to replace txid
op := types.NewOutPoint(&txid[I], vout[i])

File: rpc/rpcutil/transaction.go [hunk 30]
```diff
@@ -401,6 +404,31 @@ func MakeUnsignedTx(
 	return tx, utxos, err
 }
 
+//CreateRawTransaction create a tx without signature,it returns a tx and utxo
+func CreateRawTransaction(
+	from string, txid []crypto.HashType, vout []uint32, to []string, amounts []uint64, height uint32,
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: do not need height parameter

File: rpc/rpcutil/transaction.go [hunk 31]
```diff
@@ -423,9 +451,15 @@ func MakeUnsignedSplitAddrTx(
 func MakeUnsignedTokenIssueTx(
 	wa service.WalletAgent, issuer, owner string, tag *rpcpb.TokenTag, fee uint64,
 ) (*types.Transaction, uint32, []*rpcpb.Utxo, error) {
+	if tag.GetDecimal() < 0 || tag.GetDecimal() > maxDecimal {
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: it is impossible for a uint32 being less than 0

File: rpc/rpcutil/transaction.go [hunk 32]
```diff
@@ -423,9 +451,15 @@ func MakeUnsignedSplitAddrTx(
...
+		return nil, 0, nil, txlogic.ErrInvalidArguments
+	}
+	if uint64(tag.GetSupply()) > math.MaxUint64/uint64(math.Pow10(int(tag.GetDecimal()))) {
+		return nil, 0, nil, txlogic.ErrInvalidArguments
+	}
 	utxos, err := wa.Utxos(issuer, nil, fee)
 	if err != nil {
-		return nil, 0, nil, err
+		return nil, 0, nil, txlogic.ErrInvalidArguments
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: why change to txlogic.ErrInvalidArguments here

File: rpc/rpcutil/transaction.go [hunk 33]
```diff
@@ -423,9 +451,15 @@ func MakeUnsignedSplitAddrTx(
 func MakeUnsignedTokenIssueTx(
 	wa service.WalletAgent, issuer, owner string, tag *rpcpb.TokenTag, fee uint64,
 ) (*types.Transaction, uint32, []*rpcpb.Utxo, error) {
+	if tag.GetDecimal() < 0 || tag.GetDecimal() > maxDecimal {
+		return nil, 0, nil, txlogic.ErrInvalidArguments
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: more error details

File: rpc/rpcutil/transaction.go [hunk 34]
```diff
@@ -479,3 +513,8 @@ func calcChangeAmount(amounts []uint64, fee uint64, utxos ...*rpcpb.Utxo) (uint6
 	changeAmt := uv - total
 	return changeAmt, changeAmt > uv
 }
+func hashFromUint64(n uint64) crypto.HashType {
+	bs := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bs, n)
+	return crypto.DoubleHashH(bs)
+}
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: unknown symbol

File: rpc/rpcutil/transaction.go [hunk 35]
```diff
@@ -479,3 +513,8 @@ func calcChangeAmount(amounts []uint64, fee uint64, utxos ...*rpcpb.Utxo) (uint6
 	changeAmt := uv - total
 	return changeAmt, changeAmt > uv
 }
+func hashFromUint64(n uint64) crypto.HashType {
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: don't add unused function

File: rpc/server/rpc_transaction.go [hunk 36]
```diff
@@ -18,7 +21,7 @@ import (
 	"github.com/BOXFoundation/boxd/rpc/pb"
 	"github.com/BOXFoundation/boxd/rpc/rpcutil"
 )
-
+const maxDecimal=8
```
Comment (Jerick26) [inline rpc/server/rpc_transaction.go]: format

File: rpc/server/rpc_transaction.go [hunk 37]
```diff
@@ -174,6 +177,75 @@ func newSendTransactionResp(code int32, msg, hash string) *rpcpb.SendTransaction
...
+
+	txpool := s.server.GetTxHandler()
+	if err := txpool.ProcessTx(tx, core.BroadcastMode); err != nil {
+		return newSendTransactionResp(-1, err.Error(), ""), nil
+	}
+	hash, _ := tx.TxHash()
+	return newSendTransactionResp(0, "success", hash.String()), nil
+}
+
+func (s *txServer) CreateRawTransaction(
```
Comment (Jerick26) [inline rpc/server/rpc_transaction.go]: need to remove it

File: commands/box/wallet/root.go (hunk 24 above)
Comment (gtygo) [reply commands/box/wallet/root.go]: Behind another loop, so you need to type it in one line.

File: rpc/rpcutil/transaction.go (hunk 32 above)
Comment (gtygo) [reply rpc/rpcutil/transaction.go]: makeunsignedtx is the same error return

File: commands/box/ctl/root.go [hunk 38]
```diff
@@ -337,6 +331,99 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+		return
+	}
+	txByte, err := hex.DecodeString(args[0])
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	tx := types.Transaction{}
+	error := tx.Unmarshal(txByte)
+	if error != nil {
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: if err := tx.Unmarshal(txByte); err != nil {
    ...
}

File: commands/box/ctl/root.go [hunk 39]
```diff
@@ -337,6 +331,99 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+		fmt.Println("Invalide argument number")
+		return
+	}
+	txByte, err := hex.DecodeString(args[0])
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	tx := types.Transaction{}
+	error := tx.Unmarshal(txByte)
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: tx := new(types.Transaction)

File: commands/box/ctl/root.go [hunk 40]
```diff
@@ -337,6 +331,99 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+func createRawTransaction(cmd *cobra.Command, args []string) {
+	if len(args) < 4 {
+		fmt.Println("Invalide argument number")
+		return
+	}
+	fmt.Println("createRawTx called")
+	from := args[0]
+	//Cut characters around commas
+	txidStr := strings.Split(args[1], ",")
+	txid := make([]crypto.HashType, 0)
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: should be txHash

File: commands/box/ctl/root.go [hunk 41]
```diff
@@ -337,6 +331,99 @@ func getRawTxCmdFunc(cmd *cobra.Command, args []string) {
...
+	for _, x := range txidStr {
+		tmp := crypto.HashType{}
+		tmp.SetString(x)
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: error check

File: rpc/rpcutil/transaction.go [hunk 42]
```diff
@@ -401,6 +403,31 @@ func MakeUnsignedTx(
 	return tx, utxos, err
 }
 
+//CreateRawTransaction create a tx without signature,it returns a tx and utxo
+func CreateRawTransaction(
```
Comment (Jerick26) [inline rpc/rpcutil/transaction.go]: add usage comments

File: commands/box/ctl/root.go [hunk 43]
```diff
@@ -51,9 +51,7 @@ func init() {
 		&cobra.Command{
 			Use:   "createrawtx",
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: usage comments

File: commands/box/ctl/root.go [hunk 44]
```diff
@@ -51,9 +51,7 @@ func init() {
...
 			Short: "Create a raw transaction",
-			Run: func(cmd *cobra.Command, args []string) {
-				fmt.Println("createrawtx called")
-			},
+			Run:   createRawTransaction,
 		},
 		&cobra.Command{
 			Use:   "debuglevel [debug|info|warning|error|fatal]",
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: incorrect usage comments

File: commands/box/ctl/root.go [hunk 45]
```diff
@@ -375,14 +463,45 @@ func validateMessageCmdFunc(cmd *cobra.Command, args []string) {
...
+		return
+	}
+	defer conn.Close()
+	txByte, err := hex.DecodeString(args[0])
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	tx := new(types.Transaction)
+	error := tx.Unmarshal(txByte)
```
Comment (Jerick26) [inline commands/box/ctl/root.go]: as above

File: commands/box/token/root.go [hunk 46]
```diff
@@ -79,6 +80,14 @@ func createTokenCmdFunc(cmd *cobra.Command, args []string) {
 	tokenSymbol := args[3]
 	tokenTotalSupply, err2 := strconv.Atoi(args[4])
 	tokenDecimals, err3 := strconv.Atoi(args[5])
```
Comment (Jerick26) [inline commands/box/token/root.go]: first error check, then value check. 
add tokenSymbol check

File: commands/box/wallet/root.go [hunk 47]
```diff
@@ -155,6 +153,28 @@ func listAccountCmdFunc(cmd *cobra.Command, args []string) {
...
+		return
+	}
+	addr := args[0]
+	wltMgr, err := wallet.NewWalletManager(walletDir)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	walletinfo, exist := wltMgr.GetAccount(addr)
+	fmt.Println(exist)
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: no meaning log

File: commands/box/wallet/root.go [hunk 48]
```diff
@@ -179,3 +199,71 @@ func dumpPrivKeyCmdFunc(cmd *cobra.Command, args []string) {
...
+			passphrase, err := wallet.ReadPassphraseStdin()
+			if err != nil {
+				fmt.Println(err)
+				return
+			}
+			privateKey, err := wltMgr.DumpPrivKey(addr, passphrase)
+			if err != nil {
+				fmt.Print("Wrong password ")
+				if i == 1 {
+					fmt.Println("This wallet dump failed ")
```
Comment (Jerick26) [inline commands/box/wallet/root.go]: should return

File: rpc/server/rpc_transaction.go [hunk 49]
```diff
@@ -336,7 +360,12 @@ func (s *txServer) MakeUnsignedTokenIssueTx(
 				string(bytes), resp)
 		}
 	}()
-	//
+	if req.GetTag().GetDecimal() > maxDecimal {
+		return newMakeTokenIssueTxResp(-1, "The range of decimal must be between 0 and 8"), nil
+	}
+	if uint64(req.Tag.Supply) > math.MaxUint64/uint64(math.Pow10(int(req.Tag.Decimal))) {
```
Comment (Jerick26) [inline rpc/server/rpc_transaction.go]: no need uint64 converstion

**Pull request merged**